src/plots/plots.js

```diff
@@ -505,12 +506,38 @@ plots.supplyDefaults = function(gd) {
     // update object references in calcdata
     if((gd.calcdata || []).length === newFullData.length) {
         for(i = 0; i < newFullData.length; i++) {
-            var trace = newFullData[i];
-            (gd.calcdata[i][0] || {}).trace = trace;
+            var newTrace = newFullData[i];
+            var cd0 = gd.calcdata[i][0];
+            if(cd0 && cd0.trace) {
+                if(cd0.trace._hasCalcTransform) {
+                    remapTransformedArrays(cd0, newTrace);
+                } else {
+                    cd0.trace = newTrace;
+                }
+            }
         }
     }
 };
 
+function remapTransformedArrays(cd0, newTrace) {
+    var oldTrace = cd0.trace;
+    var arrayAttrs = PlotSchema.findArrayAttributes(oldTrace);
+    var transformedArrayHash = {};
+    var i, ast;
+
+    for(i = 0; i < arrayAttrs.length; i++) {
+        ast = arrayAttrs[i];
+        transformedArrayHash[ast] = Lib.nestedProperty(oldTrace, ast).get().slice();
+    }
+
+    cd0.trace = newTrace;
+
+    for(i = 0; i < arrayAttrs.length; i++) {
+        ast = arrayAttrs[i];
+        Lib.nestedProperty(cd0.trace, ast).set(transformedArrayHash[ast]);
```

Comment:
Here we make sure that the array attributes in `gd.calcdata[i][0].trace` match the transformed array post calc transform. Other non-data-array attributes are left untouched.